### PR TITLE
For WAFS, add warning of missing data, change cpreq to cp -v

### PR DESCRIPTION
**Pull request by YaliMao-NOAA**
Change cpreq to cp -v
Add warning of missing data outside of SENDMAIL blocks

## Pull Request Testing ##

- [ ] Describe testing already performed for this Pull Request:</br>

- [ ] Recommend testing for the reviewer(s) to perform, including the location of input datasets, and any additional instructions:</br>

- [ ] Has the code been checked to ensure that no errors occur during the execution? **[Yes or No]**

- [ ] Do these updates/additions include sufficient testing updates? **[Yes or No]**

- [ ] Please complete this pull request review by **[Fill in date]**.</br>

## Pull Request Checklist ##

- [ ] Review the source issue metadata (required labels, projects, and milestone).
- [ ] Complete the PR description above.
- [ ] Ensure the PR title matches the feature branch name.
- [ ] Check the following:
- [ ]  Instructions provided on how to run
- [ ]  Developer's name is replaced by ${user} where necessary throughout the code
- [ ]  Check that the ecf file has all the proper definitions of variables
- [ ]  Check that the jobs file has all the proper settings of COMIN and COMOUT and other input variables
- [ ]  Check to see that the output directory structure is followed
- [ ]  Be sure that you are not using MET utilities outside the METplus wrapper structure

- [ ] After submitting the PR, select **Development** issue with the original issue number.
- [ ] After the PR is approved, merge your changes. If permissions do not allow this, request that the reviewer do the merge.
- [ ] Close the linked issue.

### CODE REVIEW

**Comment by AliciaBentley-NOAA** [comment]
@YaliMao-NOAA Thank you for submitting this PR! Can you please update your new warning lines to say "WARNING" (in all caps) instead of "Warning"? After those updates are made and this PR is tested, I approve this PR to be merged! Thanks!

CC @PerryShafran-NOAA @ShelleyMelchior-NOAA

**Comment by PerryShafran-NOAA** [comment]
@YaliMao-NOAA Is this PR ready for test?  We'll need to test the missing data feature, and to do that, we might need to fake a directory in order to test that new feature.  Then we can set it back to the proper directory settings and then do another run to test the cp -v feature.

**Comment by YaliMao-NOAA** [comment]
@PerryShafran-NOAA I added warnings for each missing forecast, but have not confirmed whether an email should be sent out for each missing forecast. Other than that, this PR is ready for test.

**Comment by PerryShafran-NOAA** [comment]
@YaliMao-NOAA I am changing your J-job to make fake directories, so we can test the missing data features.  I set the following:

```
export COMIN=${COMIN:-$(compath.py ${envir}/coms/${NET}/${evs_ver})}
export DCOMINuk=${DCOMINuk:-$DCOMROOT} # $VDATE/wgrbbuls/ukmet_wafs}
export COMINgfs=${COMINgfs:-$(compath.py ${envir}/coms/gfs/${gfs_ver})}
```

That way we will see if your missing data changes are good.

**Comment by PerryShafran-NOAA** [comment]
@YaliMao-NOAA 

The missing data test has been completed.

Will you check the .o file /lfs/h2/emc/vpppg/noscrub/perry.shafran/pr440test/EVS/dev/drivers/scripts/stats/wafs/jevs_wafs_stats.o132127212 to see if we  have the WARNING statements in the correct places?

**Comment by YaliMao-NOAA** [comment]
@PerryShafran-NOAA There are false alarms of missing forecasts because icing has analysis data at every 3 hours. For example, /gfs.20240204/15/atmos/gfs.t15z.wafs_0p25_unblended.f36.grib2. I need to figure out how to avoid these.

**Comment by PerryShafran-NOAA** [comment]
@YaliMao-NOAA I'm going to do something else, and that's set up fake directories in the driver script rather than set up the J-job the way I did.  I just ran into these issues with Shannon's PR.  Please hold for a sec and let me do another run.

**Comment by PerryShafran-NOAA** [comment]
@YaliMao-NOAA I am adding these lines to your driver script:

```
export DCOMINuk=/lfs/h1/ops/pppp/dcom
export COMINgfs=/lfs/h1/ops/pppp/com/gfs/${gfs_ver}
```

These are fake directories, but it will give better warning statements than the previous job.

**Comment by PerryShafran-NOAA** [comment]
@YaliMao-NOAA The runs are complete.  Please check the latest .o file jevs_wafs_stats.o132134267.  Please note again that I am using a fake directory to test the WARNING statement capabilities.  I will run again without the fake directories, but after you check that this one is good with the WARNING statements.

**Comment by PerryShafran-NOAA** [comment]
Please do a sync fork.

**Comment by YaliMao-NOAA** [comment]
I am not able to access WCOSS2.

**Comment by PerryShafran-NOAA** [comment]
@YaliMao-NOAA Oh, are you having troubles with the connection?  It should be up and running now.

**Comment by YaliMao-NOAA** [comment]
That's weird. I just tried and got error message:
Connection open failed; No more authentication methods available (local): Authentication failed

**Comment by PerryShafran-NOAA** [comment]
You should probably contact the helpdesk for that.  I had that problem last week.

**Comment by PerryShafran-NOAA** [comment]
@YaliMao-NOAA You can also try rebooting your computer - that's what helped me solve my problem last week.

**Comment by YaliMao-NOAA** [comment]
@PerryShafran-NOAA Thank you. I rebooted the computer and it works!
I am running a test run without the false WARNINGs because GFS doesn't have 03/09/15/21z. I will get back to you once the test run is complete and good

**Comment by YaliMao-NOAA** [comment]
@PerryShafran-NOAA It's done

**Comment by PerryShafran-NOAA** [comment]
@YaliMao-NOAA Running another test of this with the fake directories and your changes pulled in.  In the meantime, please sync your fork again.

**Comment by PerryShafran-NOAA** [comment]
@YaliMao-NOAA Please examine the .o file /lfs/h2/emc/vpppg/noscrub/perry.shafran/pr440test/EVS/dev/drivers/scripts/stats/wafs/jevs_wafs_stats.o132221640 and let me know if this is OK.  Again, note that we are looking at fake directories to test the warnings output.  Then I will run again with a real run to check the cp-v output.

**Comment by PerryShafran-NOAA** [comment]
@YaliMao-NOAA I also did a real run with no fake directories, to test the cp -v output.  This particular run is here:  /lfs/h2/emc/vpppg/noscrub/perry.shafran/pr440test/EVS/dev/drivers/scripts/stats/wafs/jevs_wafs_stats.o132225469

The output is here:   /lfs/h2/emc/vpppg/noscrub/perry.shafran/evs/v1.0/stats/wafs/wafs.20240207

I notice that in the output of this run we still have many WARINGS, but they are all for ukmet wafs files, for missing files at 03Z, 09Z, 15Z, and 21Z.  You'll probably need to fix this in the same way you fixed the warnings for the GFS at those same times. 

@AliciaBentley-NOAA @ShelleyMelchior-NOAA for awareness.

File: scripts/stats/wafs/exevs_wafs_atmos_stats.sh [hunk 1]
```diff
@@ -70,8 +70,8 @@ for resolution in $resolutions ; do
 done
 
 if [ $SENDCOM = YES ] ; then
-    [ -z `ls -A $STATSOUTfinal` ] || cpreq $STATSOUTfinal/* $COMOUTfinal/.  
-    [ -z `ls -A $STATSOUTsmall` ] || cpreq $STATSOUTsmall/* $COMOUTsmall/.
+    [ -z `ls -A $STATSOUTfinal` ] || cp -v $STATSOUTfinal/* $COMOUTfinal/.
+    [ -z `ls -A $STATSOUTsmall` ] || cp -v $STATSOUTsmall/* $COMOUTsmall/.
 fi
```
Comment (PerryShafran-NOAA) [inline scripts/stats/wafs/exevs_wafs_atmos_stats.sh:76]: @YaliMao-NOAA 

I'm getting this output in the latest wafs test run:

```
626 + '[' -z evs.stats.wafs.atmos.grid2grid_icesev.v20240207.stat evs.stats.wafs.atmos.grid2grid_uvt1p25.v20240207.stat ']'
/lfs/h2/emc/vpppg/noscrub/perry.shafran/pr440test/EVS/scripts/stats/wafs/exevs_wafs_atmos_stats.sh: line 73: [: evs.stats.wafs.atmos.
grid2grid_icesev.v20240207.stat: binary operator expected
```

`/lfs/h2/emc/vpppg/noscrub/perry.shafran/pr440test/EVS/scripts/stats/wafs/exevs_wafs_atmos_stats.sh: line 74: [: too many arguments`

I'm not sure that -z is appropriate here.  I think it should be -s.  But I'm not entirely sure that these items can take multiple arguments.  I'll test that and let you know.  

@AliciaBentley-NOAA @ShelleyMelchior-NOAA for awareness.

Comment (PerryShafran-NOAA) [reply scripts/stats/wafs/exevs_wafs_atmos_stats.sh:76]: @YaliMao-NOAA This syntax doesn't work with -s either.  I'm doing a slight rewrite of this block of code to ensure that we have our file checks.  Each file must be checked individually. I will show you the code once I've tested it.

Comment (PerryShafran-NOAA) [reply scripts/stats/wafs/exevs_wafs_atmos_stats.sh:76]: @YaliMao-NOAA I did a rewrite of this script, to make sure we get rid of these errors that we saw above.  Please take this file and place it within your fork, please.  
/lfs/h2/emc/vpppg/noscrub/perry.shafran/pr440test/EVS/scripts/stats/wafs/exevs_wafs_atmos_stats.sh

Please check out the .o file that ran with this code:

/lfs/h2/emc/vpppg/noscrub/perry.shafran/pr440test/EVS/dev/drivers/scripts/stats/wafs/jevs_wafs_stats.o132229810

And compare the "saved" output directories with the directories run with this code:

/lfs/h2/emc/vpppg/noscrub/perry.shafran/evs/v1.0/stats/wafs

```
10 (dlogin08) /lfs/h2/emc/vpppg/noscrub/perry.shafran/evs/v1.0/stats/wafs > ls -ld *20240207*
drwxr-sr-x 3 perry.shafran vpppg 4096 Feb  8 14:48 atmos.20240207
drwxr-sr-x 3 perry.shafran vpppg 4096 Feb  8 14:48 atmos.20240207_save
drwxr-sr-x 2 perry.shafran vpppg 4096 Feb  8 15:49 wafs.20240207
drwxr-sr-x 2 perry.shafran vpppg 4096 Feb  8 15:49 wafs.20240207_save
11 (dlogin08) /lfs/h2/emc/vpppg/noscrub/perry.shafran/evs/v1.0/stats/wafs
```

**Comment by YaliMao-NOAA** [comment]
@PerryShafran-NOAA 
Sorry I just came back from my doctor appointment. I did another commit with emailing partial missing forecast files and with a sync.

**Comment by YaliMao-NOAA** [comment]
> I notice that in the output of this run we still have many WARINGS, but they are all for ukmet wafs files, for missing files at 03Z, 09Z, 15Z, and 21Z

Thank you for the catch. I am going to fix it.

**Comment by PerryShafran-NOAA** [comment]
@YaliMao-NOAA Thanks!

**Comment by YaliMao-NOAA** [comment]
@PerryShafran-NOAA Done.

**Comment by PerryShafran-NOAA** [comment]
@YaliMao-NOAA Can you also make one more change?  We needed to change the file checking so that file checks are done on individual files.  Please replace your exevs script with this one: 
 /lfs/h2/emc/vpppg/noscrub/perry.shafran/pr440test/EVS/scripts/stats/wafs/exevs_wafs_atmos_stats.sh
 
 Once this file is in your fork, I'll test one more time.

**Comment by YaliMao-NOAA** [comment]
@PerryShafran-NOAA 
My fix of double quoting `ls -A $STATSOUTfinal` works. It's simpler so I didn't pick up your fix. Mine is at /lfs/h2/emc/vpppg/noscrub/yali.mao/git/EVS/scripts/stats/wafs/exevs_wafs_atmos_stats.sh. The following is the difference:

<      for FILE in $STATSOUTfinal/*; do
< 	if [ -s $FILE ]; then
< 	  cp -v $FILE $COMOUTfinal/.
< 	fi
<      done
<      for FILE2 in $STATSOUTsmall/*; do
< 	if [ -s $FILE2 ]; then
< 	   cp -v $FILE2 $COMOUTsmall/.
< 	fi
<      done
---
>     [ -z "`ls -A $STATSOUTfinal`" ] || cp -v $STATSOUTfinal/* $COMOUTfinal/.
>     [ -z "`ls -A $STATSOUTsmall`" ] || cp -v $STATSOUTsmall/* $COMOUTsmall/.

**Comment by PerryShafran-NOAA** [comment]
@YaliMao-NOAA Yes, but if you notice in the subthread above, it was also causing errors in the output.  Those errors would have been flagged by NCO.  Those errors are here:

```
626 + '[' -z evs.stats.wafs.atmos.grid2grid_icesev.v20240207.stat evs.stats.wafs.atmos.grid2grid_uvt1p25.v20240207.stat ']'
/lfs/h2/emc/vpppg/noscrub/perry.shafran/pr440test/EVS/scripts/stats/wafs/exevs_wafs_atmos_stats.sh: line 73: [: evs.stats.wafs.atmos.
grid2grid_icesev.v20240207.stat: binary operator expected
```

`/lfs/h2/emc/vpppg/noscrub/perry.shafran/pr440test/EVS/scripts/stats/wafs/exevs_wafs_atmos_stats.sh: line 74: [: too many arguments`

Thus, we will need you to change to my code in order to ensure that the file checks are done and do not produce any error statements like these.  Please do so.

**Comment by YaliMao-NOAA** [comment]
@PerryShafran-NOAA With the double quoting fix, there is no such error of too many arguments.
Please check the log file, the error message is gone. /lfs/h2/emc/vpppg/noscrub/yali.mao/git/EVS/jevs_wafs_stats.o132231684

**Comment by PerryShafran-NOAA** [comment]
OK, cool.  Let me pull those in and test.  As long as we have no errors, then that should be good.

**Comment by PerryShafran-NOAA** [comment]
@YaliMao-NOAA Also please sync your fork once again.

**Comment by PerryShafran-NOAA** [comment]
@YaliMao-NOAA Your latest code updates are complete.  It looks a lot better in that there are no errors relating to the file checks and copy -v.  

Please look at the latest .o file here:  /lfs/h2/emc/vpppg/noscrub/perry.shafran/pr440test/EVS/dev/drivers/scripts/stats/wafs/jevs_wafs_stats.o132238535

There are a few WARNINGs but they look to be legit warnings, which is good and we want to keep.  

```
30 (dlogin04) /lfs/h2/emc/vpppg/noscrub/perry.shafran/pr440test/EVS/dev/drivers/scripts/stats/wafs > grep 'WARNING' jevs_wafs_stats.o132238535
WARNING: missing forecast /lfs/h1/ops/prod/com/gfs/v16.3/gfs.20240206/12/atmos/WAFS_0p25_blended_2024020612f24.grib2
0 + echo 'WARNING: missing forecast /lfs/h1/ops/prod/com/gfs/v16.3/gfs.20240206/12/atmos/WAFS_0p25_blended_2024020612f24.grib2'
WARNING: Some blend forecasts  F24 are missing for GCIP valid date 2024020712. METplus continues.
0 + echo 'WARNING: Some blend forecasts  F24 are missing for GCIP valid date 2024020712. METplus continues.'
```

Please check that, and also examine once again the output files here:  /lfs/h2/emc/vpppg/noscrub/perry.shafran/evs/v1.0/stats/wafs/____.20240207

**Comment by PerryShafran-NOAA** [comment]
@YaliMao-NOAA The message "Some blend forecasts F24 are missing" can be kind of confusing to NCO.  With the warning of the missing file above that, I'm not sure that this second warning is necessary.  Can you remove the "Some blend forecasts" warning statement entirely?  We already have statements about each missing file, and that should be enough.

**Comment by YaliMao-NOAA** [comment]
@PerryShafran-NOAA Yes, the warning is legit.
I just commented on "EVS v1.0 Issues" google document, and included you. https://docs.google.com/document/d/1VYGABeEe2eXA1k5NFL5oavgMgiiTPyTKGWk-zWmsPdc/edit#heading=h.y9yud7qhzjxm
I took both of Shelley and Wei's suggestions. I will issue an individual warning for each missing forecast, and send out email once for all missing forecast.
This is the reason why you saw the warnings twice

**Comment by PerryShafran-NOAA** [comment]
> @YaliMao-NOAA Please examine the .o file /lfs/h2/emc/vpppg/noscrub/perry.shafran/pr440test/EVS/dev/drivers/scripts/stats/wafs/jevs_wafs_stats.o132221640 and let me know if this is OK. Again, note that we are looking at fake directories to test the warnings output. Then I will run again with a real run to check the cp-v output.

Also I need to know if you have already looked into this run with the fake directories and saw the WARNING statements here?

**Comment by PerryShafran-NOAA** [comment]
> @PerryShafran-NOAA Yes, the warning is legit. I just commented on "EVS v1.0 Issues" google document, and included you. https://docs.google.com/document/d/1VYGABeEe2eXA1k5NFL5oavgMgiiTPyTKGWk-zWmsPdc/edit#heading=h.y9yud7qhzjxm I took both of Shelley and Wei's suggestions. I will issue an individual warning for each missing forecast, and send out email once for all missing forecast. This is the reason why you saw the warnings twice

OK, that does make sense.   So the 2nd warning is for the mail, the 1st is not?  It doesn't totally matter though because there will be no SENDMAIL in ops.  But perhaps remove the word "Some" and just say "Blend forecasts F24 are missing" - presumably if there are more missing forecasts, the statement would say "Blend forecasts F24, F36 are missing" or something like that?

@AliciaBentley-NOAA - do you think that the 2nd warning is OK in this case?  Wei did advise one email with a list of missing files rather than a mail for each file.

**Comment by YaliMao-NOAA** [comment]
> OK, that does make sense. So the 2nd warning is for the mail, the 1st is not?

Yes.

**Comment by YaliMao-NOAA** [comment]
> > @YaliMao-NOAA Please examine the .o file /lfs/h2/emc/vpppg/noscrub/perry.shafran/pr440test/EVS/dev/drivers/scripts/stats/wafs/jevs_wafs_stats.o132221640 and let me know if this is OK. Again, note that we are looking at fake directories to test the warnings output. Then I will run again with a real run to check the cp-v output.
> 
> Also I need to know if you have already looked into this run with the fake directories and saw the WARNING statements here?

I am seeing false alarm of UK missing data at 03 09 15 21z, for example EGRR_WAFS_0p25_icing_unblended_20240205_15z_t36.grib2. I think it has been fixed.

**Comment by PerryShafran-NOAA** [comment]
@YaliMao-NOAA Yes, that was before I pulled in your fix.  And I see that this ukmet false warning was the subject of Wei's statement 39 under item 86, so yes, your fix does address that.  So we are actually good on that front.

Can you adjust your second WARNING statement so that it is clearer?  Something like 

WARNING: A subset of blend forecasts (F24) are missing for GCIP valid date 2024020712. METplus continues.

That is clearer to people than the existing statement.  Can we make that change?

**Comment by YaliMao-NOAA** [comment]
> Can you adjust your second WARNING statement so that it is clearer?

@PerryShafran-NOAA Thank you for the suggestion. Done

File: ush/wafs/evs_wafs_atmos_stats_preparedata.sh [hunk 2]
```diff
@@ -103,12 +103,12 @@ if [ -z $FHOURS_EVSlist ] ; then
...
 	if [[ $SENDMAIL = YES ]] ; then
-            export subject="Some forecast files are missing for EVS ${COMPONENT}"
-            echo "WARNING: Some $CENTER forecasts $missingForecast are missing for $OBSERVATION valid date ${VDATE}${cc}. METplus continues." > mailmsg.$OBSERVATION.$CENTER.$RESOLUTION
+            export subject="A subset of $CENTER forecast files is missing for EVS ${COMPONENT}"
+            echo "WARNING: A subset of $CENTER forecasts ($missingForecast) is missing for $OBSERVATION valid date ${VDATE}${cc}. METplus continues." > mailmsg.$OBSERVATION.$CENTER.$RESOLUTION
             echo "Job ID: $jobid" >> mailmsg.$OBSERVATION.$CENTER.$RESOLUTION
 	    cat mailmsg.$OBSERVATION.$CENTER.$RESOLUTION | mail -s "$subject" $MAILTO
 	fi
-	echo "WARNING: Some $CENTER forecasts $missingForecast are missing for $OBSERVATION valid date ${VDATE}${cc}. METplus continues."
+	echo "WARNING: A subset of $CENTER forecasts ($missingForecast) is missing for $OBSERVATION valid date ${VDATE}${cc}. METplus continues."
```
Comment (AliciaBentley-NOAA) [inline ush/wafs/evs_wafs_atmos_stats_preparedata.sh]: This echo statement is repeated outside the SENDMAIL block. I thought the consolidated WARNING statements were only going in the SENDMAIL block. If so, and if we still have the warning statements that list the individual files that are missing, can this repeated consolidated statement be removed? @YaliMao-NOAA @PerryShafran-NOAA

**Comment by PerryShafran-NOAA** [comment]
@YaliMao-NOAA Alicia's statement suggests that the WARNING statement, that is repeated both inside and outside the SENDMAIL block, should *only* be inside the SENDMAIL block.  Please remove the WARNING statement outside the SENDMAIL block.  Thanks!

@AliciaBentley-NOAA for awareness.

**Comment by YaliMao-NOAA** [comment]
> Please remove the WARNING statement outside the SENDMAIL block

Done!

**Comment by PerryShafran-NOAA** [comment]
@YaliMao-NOAA Thank you!  One final test then.

**Comment by PerryShafran-NOAA** [comment]
@YaliMao-NOAA This is good!  I think we are now ready for merge.  Thanks for your patience with me in working on this one.

**Review by PerryShafran-NOAA — APPROVED** [review]
Tests give results as expected.  Approved for merge.

**Pull request merged**